### PR TITLE
Basic Authentication - kubeshop/kusk-gateway#399

### PR DESCRIPTION
**Pull request by mbana**
Basic Authentication - kubeshop/kusk-gateway#399

First attempt at basic auth.

See <https://github.com/kubeshop/kusk-gateway/issues/399> for further information.

---

I will remove the `Println`s and debugging statements later.

---

How to run it. Checkout the branch above and

```sh
git pull
git checkout mbana-basic-auth-prototype
make create-env
kubectl port-forward service/default 8080:80
```

```sh
kubectl apply -f protoypes/manifests/ext-authz-http-service-api.yaml
kubectl apply -f protoypes/manifests/auth-basic-staticroute.yaml
kubectl apply -f protoypes/manifests/ext-authz-http-service.yaml
```

Run the commands below. Notice the first one contains invalid credentials (`Unauthorized - hint: credentials are kubeshop:kubeshop`) causing a `401 Unauthorized`, then we try again with valid credentials, `kubeshop:kubeshop`, and then it works.

```sh
$ curl -L -X GET -v --user kubeshop:kubeshopx http://localhost:8080/get2
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
* Server auth using Basic with user 'kubeshop'
> GET /get2 HTTP/1.1
> Host: localhost:8080
> Authorization: Basic a3ViZXNob3A6a3ViZXNob3B4
> User-Agent: curl/7.83.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< x-powered-by: Express
< date: Wed, 01 Jun 2022 12:56:11 GMT
< content-length: 54
< x-envoy-upstream-service-time: 761
< content-type: text/plain
< server: envoy
<
* Connection #0 to host localhost left intact
Unauthorized - hint: credentials are kubeshop:kubeshop%
$ curl -L -X GET -v --user kubeshop:kubeshop http://localhost:8080/get2
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
* Server auth using Basic with user 'kubeshop'
> GET /get2 HTTP/1.1
> Host: localhost:8080
> Authorization: Basic a3ViZXNob3A6a3ViZXNob3A=
> User-Agent: curl/7.83.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< server: envoy
< date: Wed, 01 Jun 2022 12:56:17 GMT
< content-type: application/json
< content-length: 367
< access-control-allow-origin: *
< access-control-allow-credentials: true
< x-envoy-upstream-service-time: 32
<
{
  "args": {},
  "headers": {
    "Accept": "*/*",
    "Authorization": "Basic a3ViZXNob3A6a3ViZXNob3A=",
    "Host": "localhost:8080",
    "User-Agent": "curl/7.83.0",
    "X-Current-User": "kubeshop",
    "X-Envoy-Expected-Rq-Timeout-Ms": "15000",
    "X-Envoy-Original-Path": "/get2"
  },
  "origin": "172.17.0.1",
  "url": "http://localhost:8080/get"
}
* Connection #0 to host localhost left intact
```

---

This PR...

## Changes

-

## Fixes

-

## Checklist

- [ ] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

File: internal/controllers/config_manager.go
Comment (jasmingacic) [inline internal/controllers/config_manager.go:115]: I'm not sure if this is going to break anything else?
Maybe there is another way of injecting this. How can we detect that ext_authz filter is required before we do this here. If i'm not mistaken you only need httpConnectionManagerBuilder for this 
```	httpConnectionManagerBuilder.AppendFilterHTTPExternalAuthorizationFilterToStart(httpExternalAuthorizationFilter)
```

File: internal/controllers/parser.go
Comment (jasmingacic) [inline internal/controllers/parser.go:338]: I already addressed this above. Is there another way to add ext_auth filter?

File: internal/envoy/config/filters.go
Comment (jasmingacic) [inline internal/envoy/config/filters.go:49]: make sure to clean up once you are done

File: internal/envoy/config/filters.go
Comment (jasmingacic) [inline internal/envoy/config/filters.go:89]: make sure to clean up once you are done

**Review by jasmingacic — CHANGES_REQUESTED** [review]
Besides some comments it is looking good. 

Before we make any move towards merging it please remove directory `prototypes` from the PR so we don't drag this around.

File: internal/controllers/parser.go
Comment (povilasv) [inline internal/controllers/parser.go:340]: could you use the logging library the project is using instead of fmt printf?

File: internal/controllers/parser.go
Comment (povilasv) [inline internal/controllers/parser.go:341]: get rid of unneeded printfs please

File: internal/controllers/envoyfleet_resources.go
Comment (mbana) [reply internal/controllers/envoyfleet_resources.go:165]: Resolved.

Now it just contains (which I think is a useful default):

```go
			"envoy -c /etc/envoy/envoy.yaml --component-log-level filter:trace,ext_authz:trace --service-node $POD_NAME",
```

File: internal/controllers/parser.go
Comment (mbana) [reply internal/controllers/parser.go]: `HTTPExternalAuthorization` needs to come before `CORS` and `Router` otherwise `envoy` will throw an error.

There is also no point in adding the `HTTPExternalAuthorization` filter if the `auth` section is not defined by the user.

File: internal/envoy/config/filters.go
Comment (mbana) [reply internal/envoy/config/filters.go]: Resolved.

Cleaned up the code.

File: internal/envoy/config/filters.go
Comment (mbana) [reply internal/envoy/config/filters.go]: Resolved.

Cleaned up the code.

File: pkg/options/auth.go
Comment (mbana) [reply pkg/options/auth.go]: Resolved.

Implemented validation logic for `AuthOptions`.

File: internal/controllers/parser.go
Comment (mbana) [reply internal/controllers/parser.go]: Resolved.

Removed `Printf`s etc. and debugging lines from code.

File: internal/controllers/parser.go
Comment (mbana) [reply internal/controllers/parser.go]: Resolved.

Removed `Printf`s etc. and debugging lines from code.

File: examples/ext_authz/ext-authz-http-service.yaml
Comment (jasmingacic) [inline examples/ext_authz/ext-authz-http-service.yaml:19]: Maybe we should host this image in Kubeshop docker hub?

File: examples/ext_authz/ext-authz-staticroute.yaml
Comment (jasmingacic) [inline examples/ext_authz/ext-authz-staticroute.yaml:9]: can you remove the unused bit

**Review by jasmingacic — CHANGES_REQUESTED** [review]
Looking good.

Just a little bit of clean up and we are can merge it

File: examples/ext_authz/ext-authz-staticroute.yaml
Comment (mbana) [reply examples/ext_authz/ext-authz-staticroute.yaml]: Resolved.

Remove unused config.

**Review by jasmingacic — DISMISSED** [review]
looking good

**Review by povilasv — DISMISSED** [review]

File: internal/controllers/parser.go
Comment (mbana) [reply internal/controllers/parser.go]: Feel free to mark as resolved, if you feel the previous response is enough.

File: examples/ext_authz/ext-authz-http-service.yaml
Comment (mbana) [reply examples/ext_authz/ext-authz-http-service.yaml:19]: Do you want to push the image? I don’t know if I have push rights to docker.io (<https://hub.docker.com/u/kubeshop>).

Suggested image name: `kubeshop/kusk-ext-authz-http-service:latest` and `kubeshop/kusk-ext-authz-http-service:v0.0.1`

Shall I also commit the `Dockerfile` for generating the `kusk-ext-authz-http-service` image.

Comment (olensmar) [reply examples/ext_authz/ext-authz-http-service.yaml]: @mbana what's your DockerHub id? I can add you there so you can push...

File: examples/ext-authz/ext-authz-http-service-api.yaml
Comment (olensmar) [inline examples/ext-authz/ext-authz-http-service-api.yaml:20]: so will this be added as a securityScheme to the post-processed OpenAPI definition?

**Review by jasmingacic — APPROVED** [review]

File: examples/ext-authz/ext-authz-http-service-api.yaml
Comment (jasmingacic) [reply examples/ext-authz/ext-authz-http-service-api.yaml:20]: There were several catches with that. 
1. Auth can be enabled in combination with 1 or more securitySchemes at the same time which would be introducing additional complexity to our setup
2. request header based securitySchemes (basic, apikey, bearer) don't have additional field that would define endpoint and similar
3. We wanted to ship this soon so we opted for the simplest solution

I'd be happy to discuss this further into details as we said this would be the first iteration of auth.

Let us see how this will stick for now.

Comment (olensmar) [reply examples/ext-authz/ext-authz-http-service-api.yaml:20]: thanks for clarifying @jasmingacic - i think it's a balance between what makes most sense from an API builder/consumer PoV and what is easiest for us - with this approach we should definitely provide a means to get the post-processed OpenAPI definition via an endpoint - otherwise the builder would have to add both our extension **and** the corresponding securitySchema to provide the API consumer with a valid description of the API... 
@crjones thoughts?

**Comment by jasmingacic** [comment]
@povilasv @kylehodgetts can you take a look so we can wrap this up

@mbana once #454 is merged can you please add smoke tests for this

**Review by jasmingacic — APPROVED** [review]
go ahead an merge it
Also make sure that you add smoketests once #454 is merged

File: smoketests/ext-authz/ext-authz_test.go
Comment (kylehodgetts) [inline smoketests/ext-authz/ext-authz_test.go:19]: remove the commented out imports if they're not needed

File: smoketests/ext-authz/ext-authz_test.go
Comment (kylehodgetts) [inline smoketests/ext-authz/ext-authz_test.go:241]: The test is called `Test_ExtAuthz_Authorized` but we are checking for a 403?

**Review by kylehodgetts — CHANGES_REQUESTED** [review]
Looks pretty good so far, just a couple of things that need addressing

**Comment by jasmingacic** [comment]
Before merging please squash commits to a single commit or maybe two

**Review by kylehodgetts — DISMISSED** [review]

**Comment by kylehodgetts** [comment]
@jasmingacic you can squash on merge in Github so @mbana doesn't need to manually squash

**Review by jasmingacic — APPROVED** [review]

**Comment by jasmingacic** [comment]
this feature is broken

```
# kubectl apply -f examples/ext-authz/
api.gateway.kusk.io/ext-authz-http-service created
deployment.apps/ext-authz-http-service unchanged
service/ext-authz-http-service unchanged
# kubectl apply -f smoketests/samples/hello-world/mock-api.yaml 
api.gateway.kusk.io/mock-api created
# curl -v 127.0.0.1/hello
*   Trying 127.0.0.1:80...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> GET /hello HTTP/1.1
> Host: 127.0.0.1
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 403 Forbidden
< date: Fri, 17 Jun 2022 00:09:01 GMT
< server: envoy
< content-length: 0
< 
* Connection #0 to host 127.0.0.1 left intact

```

[mock-api.yaml](https://github.com/kubeshop/kusk-gateway/blob/main/smoketests/samples/hello-world/mock-api.yaml) doesn't have auth enabled an it is failing because another API has it enabled. This is misconfigured or we missed the point on how ext-auth filters are working with envoy.

Everyone please take it for a spin and get me results of your tests.

**Pull request closed without merging**
